File: CS313/Asn2/PSIM/sim.py
import random
import collections


class Params(object):
    def __init__(self, protocol, num_stations, prob, num_slots, num_trials):
        self.protocol = protocol
        self.num_stations = num_stations
        self.prob = prob
        self.num_slots = num_slots
        self.num_trials = num_trials

class Station(object):
    def __init__(self, num_stations, prob, number):
        self.num_stations = num_stations
        self.prob = prob
        self.number = number
        self.frames_sent = 0
        self.delay = 0
        self.que = collections.deque()
        self.undelivered_ratios = []
        self.trial_frames_sent = []
        self.trial_delays = []
    
    def reset(self):
        self.que = collections.deque()
        self.frames_sent = 0
        self.delay = 0
    
    def do_slot(self, attempting):
        only_frame = False
        if get_random() < self.prob:
            only_frame = self.create_frame()
        for frame in range(len(self.que)):
            self.que[frame] += 1
        will_attempt = self.check_attempt(only_frame)
        if self.que and will_attempt:
            attempting.append(self)
        
    def create_frame(self):
        if self.que: only_frame = False
        else: only_frame = True
        self.que.append(0)
        return only_frame
    
    def check_attempt(self, not_needed):
        if self.counter:
            self.counter -= 1
            return False
        else:
            return True
    
    def transmit(self):
        next_frame_delay = self.get_frame()
        self.delay += next_frame_delay
        self.frames_sent += 1
        
    def get_frame(self):
        return self.que.popleft()
    
    def store_trial(self):
        self.trial_frames_sent.append(self.frames_sent)
        self.trial_delays.append(self.delay)
        undelivered = len(self.que)
        try:
            self.undelivered_ratios.append(undelivered / float(undelivered + self.frames_sent))
        except ZeroDivisionError:
            self.undelivered_ratios.append(0.0)
    
    def get_throughput(self, total_slots):
        self.throughput = sum(self.trial_frames_sent) / float(total_slots)
    
    def get_delay(self):
        try:
            self.avg_delay = sum(self.delay) / float(self.frames_sent)
        except ZeroDivisionError: 
            self.avg_delay = 0.0
    
    def get_results(self, total_slots):
        total_frames = sum(self.trial_frames_sent)
        throughput = total_frames / float(total_slots)
        try:
            avg_delay = sum(self.trial_delays) / float(total_frames)
        except ZeroDivisionError: 
            avg_delay = 0.0
        return concat_seq([str(result) for result in [self.number, throughput, avg_delay, concat_seq(self.undelivered_ratios)]])
        
    
class P_Station(Station):
    def __init__(self, num_stations, prob, number):
        Station.__init__(self, num_stations, prob, number)
        
    def check_attempt(self, only_frame):
        if only_frame or get_random() < 1 / float(self.num_stations): 
            return True
        else: 
            return False
    
    def collide(self):
        return

class I_Station(Station):
    def __init__(self, num_stations, prob, number):
        Station.__init__(self, num_stations, prob, number)
        self.counter = 0

    def reset(self):
        Station.reset(self)
        self.counter = 0

    def collide(self):
        self.counter = random.randint(0, self.num_stations - 1)
    
class B_Station(Station):
    def __init__(self, num_stations, prob, number):
        Station.__init__(self, num_stations, prob, number)
        self.counter = 0
        self.collisions = 0

    def reset(self):
        Station.reset(self)
        self.counter = 0
        self.collisions = 0
            
    def collide(self):
        if self.collisions < 10: self.collisions += 1
        self.counter = random.randint(0, (2 ** self.collisions) - 1)
    
    def transmit(self):
        Station.transmit(self)
        self.collisions = 0

class T_Station(Station):
    def __init__(self, num_stations, prob, number):
        Station.__init__(self, num_stations, prob, number)
        self.counter = number
    
    def reset(self):
        Station.reset(self)
        self.counter = self.number

    def check_attempt(self, not_needed):
        if self.counter:
            self.counter -= 1
            return False
        else:
            self.counter = self.num_stations - 1
            return True    


# start
def run(protocol, num_stations, prob, num_slots, num_trials, seeds):
    params = Params(protocol, num_stations, prob, num_slots, num_trials)
    if protocol == 'P':
        stations = [P_Station(num_stations, prob, n) for n in range(num_stations)]
    elif protocol == 'I':
        stations = [I_Station(num_stations, prob, n) for n in range(num_stations)]
    elif protocol == 'B':
        stations = [B_Station(num_stations, prob, n) for n in range(num_stations)]
    elif protocol == 'T':
        stations = [T_Station(num_stations, prob, n) for n in range(num_stations)]
    else:
        print 'unknown protocol'
        exit()
    for trial in range(num_trials):
        run_trial(stations, params, seeds[trial])
        store_trial(stations)
    return stations

def run_trial(stations, params, seed):
    random.seed(seed)
    for station in stations: station.reset()
    for slot in range(params.num_slots):
        attempting = list()
        for station in stations:
            station.do_slot(attempting)
        transmit(attempting)

def transmit(attempting):
    tries = len(attempting)
    if not tries: return
    elif tries == 1: attempting[0].transmit()
    else:
        for station in attempting:
            try:
                station.collide()
            except AttributeError:
                print 'num stations:', tries, ', this:', station.number 

def store_trial(stations):
    for station in stations: station.store_trial()

def get_random():
    return random.random()

def concat_seq(seq):
    return ' '.join([str(item) for item in seq])

File: CS313/Asn2/PSIM/sim.c

#include "sim.h"
#include "que.c"

struct Param param;

struct Station* get_stations(int num_stations) {
	struct Station* stations[num_stations];
	for (int n = 0; n < num_stations; n++)
		stations[n] = malloc(sizeof(struct Station));
	return stations[0];
}

// start simulation
void run(char* protocol, int num_stations, float prob, 
					int num_slots, int num_trials) {

	struct Station* stations = get_stations(num_stations);
	param.protocol = *protocol; printf("%c\n", param.protocol);
	param.num_stations = num_stations; printf("%i\n", param.num_stations);
	param.prob = prob; printf("%f\n", param.prob);
	param.num_slots = num_slots; printf("%i\n", param.num_slots);
	param.num_trials = num_trials; printf("%i\n", param.num_trials);
	
	// determine station behaviour based on protocol
	if (*protocol == 'P') {
		param.check_attempt = P_check_attempt;
		param.collide = P_collide; 
	}
	else if (*protocol == 'I') {
		param.check_attempt = default_check_attempt;
		param.collide = I_collide; 
	}
	else if (*protocol == 'B') {
		param.check_attempt = default_check_attempt;
		param.collide = B_collide; 
	}
	else if (*protocol == 'T') {
		param.check_attempt = T_check_attempt;
	}
	
	// initialize stations
	for (int n = 0; n < num_stations; n++) {
		stations[n].protocol = *protocol; 
		stations[n].num_stations = num_stations; 
		stations[n].prob = prob; 
		stations[n].number = n; 
		stations[n].frames_sent = 0; 
		stations[n].delay = 0; 
		stations[n].collisions = 0;
		if (*protocol == 'T')
			stations[n].counter = n;
		else
			stations[n].counter = 0;
	}

	// run trials
    for (int trial = 0; trial < num_trials; trial++) {
        run_trial(param, stations);
		printf("finished trial %d\n", trial);
        //store_trial(stations);
    //produce_results(stations, param);
	}
    //return stations;
}

// done
void run_trial(struct Param param, struct Station *stations) {
	struct Station* attempting[param.num_stations];
	for (int station = 0; station < param.num_stations; station++) {
		stations[station].que = create_que();
	}
	puts("finished creating ques");
    for (int slot = 0; slot < param.num_slots; slot++) {
        for (int station = 0; station < param.num_stations; station++) {
            do_slot(param, stations[station], attempting);
			
		}
		puts("finished updating stations");
        transmit_frame(param, attempting);
		puts("finished transmitting");
	}
}

// done
void do_slot(struct Param param, struct Station station, struct Station* attempting[]) {
	int must_attempt = 0;
	if (get_random() < station.prob) {
		must_attempt = create_frame(&station);
	}
	increment_delays(station.que);
	int will_attempt = (param.check_attempt) (&station);
	if (!is_empty(station.que) && (must_attempt || will_attempt))
		attempting[station.number] = &station;
	else
		attempting[station.number] = NULL;
}

// done
int create_frame(struct Station* station) {
	//puts("creating frame");
	struct node *node;
	node = create_node(1);
	int must_attempt;
	if (is_empty(station->que))
		must_attempt = 1;
	else
		must_attempt = 0;
	enqueue(station->que, node);
	return must_attempt;
}

// done
void increment_delays(struct que* que) {
	for (struct node* node = que->head; node != NULL; node = node->next)
		node->value += 1;
}

// done
void transmit_frame(struct Param param, struct Station* attempting[]) {
	int num_attempting = get_attempting(param, attempting);
	printf("%d stations wish to transmit\n", num_attempting);
	if (num_attempting == 1)
		transmit(get_station(param, attempting));
	else if (num_attempting > 1) {
		for (int n = 0; n < param.num_stations; n++) {
			if (attempting[n] != NULL)
				param.collide(attempting[n]);
		}
	}
}

// done
int get_attempting(struct Param param, struct Station* attempting[]) {
	int found = 0;
	for (int n = 0; n < param.num_stations; n++) {
		if (attempting[n] != NULL)
			found += 1;
	}
	return found;
}

// done
struct Station* get_station(struct Param param, struct Station* attempting[]) {
	for (int n = 0; n < param.num_stations; n++) {
		if (attempting[n] != NULL)
			return attempting[n];
	}
	return NULL;
}

// done
int P_check_attempt(struct Station* station) {
	//puts("P check");
	if (get_random() < 1.00 / station->num_stations)
		return 1;
	else
		return 0;
}

// done
int T_check_attempt(struct Station* station) {
	//puts("T check");
	if (station->counter) {
		station->counter -= 1;
		return 0;
	}
	else {
		station->counter = station->num_stations - 1;
		return 1;
	}
}

// done
int default_check_attempt(struct Station* station) {
	//puts("default check");
	if (station->counter) {
		station->counter -= 1;
		return 0;
	}
	else {
		return 1;
	}
}

// done
void P_collide(struct Station* station) {
	;
}

// done
void I_collide(struct Station* station) {
	station->counter = get_int(station->num_stations - 1);
}

// done
void B_collide(struct Station* station) {
	if (station->collisions < 10)
		station->collisions += 1;
	station->counter = get_int(pow(2.0, (double) station->collisions) - 1);
}

// done
void transmit(struct Station* station) {
	//puts("transmit");
	station->delay += dequeue(station->que);
	station->frames_sent += 1;
	station->collisions = 0;
}

double get_random() {
	double num = drand48(); printf("random number is %lf\n", num); return num;
	//return drand48();
}

int get_int(int max_int) {
	return  (int) floor( (max_int + 1)  * get_random());
}